indico/cli/python_upgrader.py

```diff
@@ -0,0 +1,140 @@
+            clear=clear,
+            symlinks=True,
+            upgrade=upgrade,
+            with_pip=True,
+            upgrade_deps=True
+        )
+        builder.create(self.venv_path)
+        subprocess.run([str(self.venv_path / 'bin' / 'pip'), 'install', '-U', 'wheel'], check=True)
+
+    def reinstall_uwsgi(self):
+        if not self.has_uwsgi:
+            print('No uWSGI installed; reinstall not needed')
+            return
+        print('Rebuilding uWSGI for the current python version')
+        subprocess.run([str(self.venv_path / 'bin' / 'pip'), 'install', '--force-reinstall', '--no-cache', 'uwsgi'],
+                       check=True)
+
+    def _prepare(self):
+        pyenv_root = Path('~/.pyenv/').expanduser()
+        pyenv_shim = pyenv_root / 'shims/python'
+
+        if not sys.executable.startswith(str(pyenv_root)):
+            # we can't run inside a virtualenv as this would result in some of the
+            # symlinks pointing there instead of the actual pyenv python binaries
+            if os.environ.get('_PYTHON_UPGRADER_REEXEC'):
+                print(f'Still not executed via venv ({sys.executable}); aborting')
+                sys.exit(1)
+            print(f'Not executed via pyenv (probably inside virtualenv); re-executing with {pyenv_shim}')
+            os.environ['_PYTHON_UPGRADER_REEXEC'] = '1'
+            os.execl(pyenv_shim, pyenv_shim, *sys.argv)
+
+        pyenv_global_python_version = Path('~/.pyenv/version').expanduser().read_text().strip()
+        pyenv_local_python_version = subprocess.run(['pyenv', 'version-name'], capture_output=True,
+                                                    encoding='utf-8').stdout.strip()
+        if pyenv_global_python_version != pyenv_local_python_version:
+            print(f'Warning: global pyenv version: {pyenv_global_python_version}; '
+                  f'local version: {pyenv_local_python_version}')
+        print(f'Python version: {".".join(map(str, sys.version_info[:3]))}; wanted: {pyenv_local_python_version}')
+        print(f'Venv path: {self.venv_path}')
+        config_file = self.venv_path / 'pyvenv.cfg'
+        if not self.venv_path.exists() or not config_file.exists():
+            print('Venv not found, creating new one')
+            upgrade = False
+            clear = True
+        else:
+            config = self._parse_venv_config(config_file)
+            print(f'Venv found; python version {config["version"]} from {config["home"]}')
+            if config['version'] == pyenv_local_python_version:
+                print('Venv is up to date; not doing anything')
+                sys.exit(0)
+            if tuple(map(int, config['version'].split('.')[:2])) != sys.version_info[:2]:
+                print('Venv is using a different Python 3.x version; recreating')
+                clear = True
+                upgrade = False
+            else:
+                clear = False
+                upgrade = True
```

Comment:
```suggestion
        config_file = self.venv_path / 'pyvenv.cfg'
        upgrade, clear = False, True
        if not self.venv_path.exists() or not config_file.exists():
            print('Venv not found, creating new one')
        else:
            config = self._parse_venv_config(config_file)
            print(f'Venv found; python version {config["version"]} from {config["home"]}')
            if config['version'] == pyenv_local_python_version:
                print('Venv is up to date; not doing anything')
                sys.exit(0)
            if tuple(map(int, config['version'].split('.')[:2])) == sys.version_info[:2]:
                clear = False
                upgrade = True
            else:
                print('Venv is using a different Python 3.x version; recreating')
```

Reply:
I prefer being explicit here - that way, if a branch isn't handled, it will fail rather than destroy (clear=True) the venv